src/ray/raylet/worker_pool.cc

```diff
@@ -919,6 +914,13 @@ void WorkerPool::TryKillingIdleWorkers() {
         can_be_killed = false;
         break;
       }
+
+      // Skip killing the worker process if there's any inflight `Exit` RPC requests to
```

Comment:
I am still not 100% sure how moving this logic here would fix the issue. Is the problem that having this block up there only counts 1 worker per process scenario whereas here can take into account of multiple workers scenario?

Reply:
The issue only happens to Java workers because a Java worker process may include multiple workers. If the number of workers in a worker process is 1, then this code change won't make any difference.

For the multi-worker scenario in Java, please consider this case:

1) A Java worker process with 2 (for simplicity) workers named A and B is considered idle by the worker pool and is about to be killed.
2) The worker pool adds workers A and B to `pending_exit_idle_workers_` and sends `Exit` RPC request to them.
3) Worker A rejects killing and is removed from `pending_exit_idle_workers_`.
4) Somehow worker B didn't reply before the next round of `TryKillingIdleWorkers`.
5) The next round of `TryKillingIdleWorkers` starts.
6) For worker A, the "if" check at line 883 (old) failed because worker A is not in `pending_exit_idle_workers_`. The worker pool considers all workers in the worker process idle and tries to kill worker A and worker B.
6.1) Worker A is added to `pending_exit_idle_workers_` again. An `Exit` RPC request is sent to worker A.
6.2) Worker B is already in `pending_exit_idle_workers_` so the emplace doesn't change anything. An `Exit` RPC request is sent to worker B.
7) Worker A rejects killing and is removed from `pending_exit_idle_workers_`.
8) Now the reply of RPC request sent to worker B at step 2) is received: worker B rejects killing and is removed from `pending_exit_idle_workers_`.
9) The reply of RPC request sent to worker B at step 6.2) is received: worker B rejects killing and **RAY_CHECK failed when removing worker B from `pending_exit_idle_workers_` because it's not in it anymore**.

To fix the potential double removing in edge cases, I moved the `if (pending_exit_idle_workers_.count(worker->WorkerId()))` check into `for (const auto &worker : workers_in_the_same_process)` to skip killing if there's any inflight `Exit` RPC request to this worker process.